src/main/java/com/google/api/gax/grpc/InstantiatingChannelProvider.java

```
@@ -0,0 +1,308 @@
+import io.grpc.auth.ClientAuthInterceptor;
+import io.grpc.netty.NegotiationType;
+import io.grpc.netty.NettyChannelBuilder;
+import java.io.IOException;
+import java.util.List;
+import java.util.concurrent.Executor;
+
+public class InstantiatingChannelProvider implements ChannelProvider {
+  private static final String DEFAULT_GAX_VERSION = "0.1.0";
+
+  private final ExecutorProvider executorProvider;
+  private final CredentialsProvider credentialsProvider;
+  private final String serviceAddress;
+  private final int port;
+  private final String clientLibName;
+  private final String clientLibVersion;
+  private final String serviceGeneratorName;
+  private final String serviceGeneratorVersion;
+
+  private InstantiatingChannelProvider(
+      ExecutorProvider executorProvider,
+      CredentialsProvider credentialsProvider,
+      String serviceAddress,
+      int port,
+      String clientLibName,
+      String clientLibVersion,
+      String serviceGeneratorName,
+      String serviceGeneratorVersion) {
+    this.executorProvider = executorProvider;
+    this.credentialsProvider = credentialsProvider;
+    this.serviceAddress = serviceAddress;
+    this.port = port;
+    this.clientLibName = clientLibName;
+    this.clientLibVersion = clientLibVersion;
+    this.serviceGeneratorName = serviceGeneratorName;
+    this.serviceGeneratorVersion = serviceGeneratorVersion;
+  }
+
+  @Override
+  public boolean needsExecutor() {
+    return executorProvider == null;
+  }
+
+  @Override
+  public ManagedChannel getChannel() throws IOException {
+    if (needsExecutor()) {
+      throw new IllegalStateException("getChannel() called when needsExecutor() is true");
+    } else {
+      return createChannel(executorProvider.getExecutor());
+    }
+  }
+
+  @Override
+  public ManagedChannel getChannel(Executor executor) throws IOException {
+    if (!needsExecutor()) {
+      throw new IllegalStateException("getChannel(Executor) called when needsExecutor() is false");
```

Shall we throw an exception or we should allow user to override the executor here